Add predict methods

vspinu opened the pull request:
I love this package because it's fills the bill for all clustering methods in one place. But the interface and UI is rather non-Rish. 

As a starter I am adding `predict` methods for cluster objects. For this I

  - have streamlined class names (preserving the old ones for compatibility). Classes with spaces (like `cluster medoids silhouette`) are really non-standard and are cumbersome to work with when defining generics.
  - removed classes from predictor output. Mostly because those are same classes as for the actual fitted cluster objects. Same class should not be used to represent different data structures. 

If these changes are ok with you I will add tests, update vignette etc. Also, time permitting, will add print, plot and summary methods to make the UI of the package more user friendly for the seasoned R users.

mlampros commented:
@vspinu thanks for the contribution. give some time to review the changes

mlampros commented on R/clustering_functions.R:
```diff
@@ -1250,11 +1272,26 @@ predict_Medoids = function(data, MEDOIDS = NULL, distance_metric = 'euclidean',
+  out <- predict_Medoids(newdata, MEDOIDS = object$medoids,
+                         distance_metric = object$distance_metric,
+                         fuzzy = fuzzy, threads = threads)
+  if (fuzzy)
```
in the *predict_Medoids()* function when the user specifies *fuzzy = TRUE* then the function returns 3 objects:

- clusters
- fuzzy_clusters
- dissimilarity

Is your intention to follow the same output format of the *predict()* function as is the case with other algorithms that return either a vector or a matrix (in case of binary classification for instance)?

mlampros replied:
ok, then no need of an extra parameter. I thought initially that the previous predict functions will be deprecated. Thus, it's fine the 'predict.MedoidsCluster()' function.

mlampros commented:
@vspinu thanks again for the additions and the corrections (regarding the classes). I'll be glad if you add also tests and update the vignette. It will be also nice to have the print, plot and summary methods. 
I'd rather prefer to keep the old predict functions than adding deprecation messages. I think keeping the old functions is also necessary for the users that already use these functions in their workflows or packages.
Once this PR is completed please mention your necessary personal information (vorname, lastname) so that I can add you as a contributor to the package DESCRIPTION file. thanks again.

vspinu commented:
> I'll be glad if you add also tests and update the vignette. It will be also nice to have the print, plot and summary methods.

Ok. I will spend some time on this. 

> I'd rather prefer to keep the old predict functions than adding deprecation messages.

Of course. I didn't mean to deprecate those.

mlampros commented:
I've added a message so that the PR is not closed from the github-actions-bot till the changes are applied (tests, etc.)

vspinu commented:
Sorry for stalling this. Things started piling on my weekends suddenly, will try to resume this week.

vspinu commented:
I have re-added the comment, added tests and adjusted the vignette to use `predict()` method. Also added `print()` methods. The objects are now printed succinctly as follows:

```
> gmm
GMM Cluster
 Call: GMM(data = dat, gaussian_comps = 3) 
 Data cols: 42 
 Centroids: 3 
> km
KMeans Cluster
 Call: KMeans_rcpp(data = dat, clusters = clust, num_init = 5, max_iters = 100,      initializer = "optimal_init") 
 Data cols: 42 
 Centroids: 2 
 BSS/SS: 0.2800772 
 SS: 1312201 = 944683.6 (WSS) + 367517.6 (BSS)
> cm
Medoids Cluster
 Call: Clara_Medoids(data = dat, clusters = 2, samples = 5, sample_size = 0.2,      swap_phase = TRUE, fuzzy = T) 
 Data cols: 42 
 Medoids: 2 
 Cluster stats:
                                  2          3
   index                        278         34
   number_obs                   226        174
   max_dissimilarity      91.879999 114.495518
   average_dissimilarity  36.482054  67.212241
   isolation               1.396992   1.740851
```

You can modify those as you see fit.  

One problem is `Kmeans_arma` which returns only [centroids](https://github.com/mlampros/ClusterR/blob/e51a39f1ad3c84fad879616d783040d3beaabaf1/R/clustering_functions.R#L407) instead of the complete structure as returned by [`KMeans_rcpp`](https://github.com/mlampros/ClusterR/blob/e51a39f1ad3c84fad879616d783040d3beaabaf1/R/clustering_functions.R#L487). So the returned object is not classed and I have added a FIXME comment there. 

Regarding the UI, for consistency and simplicity, I would propose to actually have only 3 top level functions: `KMeans`, `Medoids` and `GMM` instead of the current (`KMeans_rcpp`, `Kmeans_arma`, `Cluster_Medoids`, `Clara_Medoids` and `GMM`). `KMeans` would have a method argument `"rcpp"` or `"arma"`, `Medoids()` would have `method` argument `"pam"` or `"clara"`.  WDYT?

mlampros commented:
First of all thanks for all your implementations. Let me see the changes in the next days. 

Regarding the UI, I think it's a good idea but we have to add deprecation messages for the previous functions so that the users can get accustomed to the new functions.

mlampros commented:
I've added a message so that this PR is not closed.

mlampros reviewed (approved):
thanks

mlampros commented:
@vspinu,

can you also update the [NEWS.md](https://cran.r-project.org/web/packages/ClusterR/news/news.html) file by adding your changes under version *1.2.6*?

vspinu commented:
Done! I have also bumped the dev version.

mlampros commented:
@vspinu,
specific tests fail. are the failures related with your changes? It seems that the functions of the failed tests require a matrix.

mlampros commented:
thank you @vspinu I'll have a look today in the afternoon

mlampros commented:
I submitted the updated version to CRAN (1.2.6). This was actually a problem with the R version I used. After updating to 4.1.1 the issue was resolved.

vspinu commented:
Perfectos! Happy to hear it solved by itself :)
